### PR TITLE
i123643 print java version gradle

### LINKED ISSUE #12357: libtxt: Text metrics are sometimes wrong (in different ways on different platforms at different sizes)
When the textScaleFactor in a RenderParagraph is not 1.0, or the font size is particular sizes (even without scaling) the Linux tests render, or at least measure, widths that are different than the same tests on Mac.  When they are off, they appear to be exactly one pixel off, and only in width.

Mac: fontSize: 10.0, textScaleFactor: 1.3 --> 13.0 wide
Linux: fontSize: 10.0, textScaleFactor: 1.3 --> 14.0 wide

Mac: fontSize: 123.0, textScaleFactor: 1.3 --> 160.0 wide
Linux: fontSize: 123.0, textScaleFactor: 1.3 --> 161.0 wide

Mac: fontSize: 100.0, textScaleFactor: 2.0 --> 200.0 wide
Linux: fontSize: 100.0, textScaleFactor: 2.0 --> 201.0 wide

BUT:
Mac: fontSize: 200.0, textScaleFactor: 2.0 --> 400.0 wide
Linux: fontSize: 200.0, textScaleFactor: 2.0 --> 400.0 wide

So it's not ALL the time, for all scales/font sizes.  This suggests some subtle difference in font rendering on one/both platforms.

Since Blink-based text rendering is being replaced, re-evaluate this bug once that has completed.

Some tests affected by this:
```
test/painting/text_painter_test.dart
test/widgets/text_test.dart
test/rendering/paragraph_test.dart
```
There is a `TODO(#12357)` in `paragraph_test.dart` that marks where the test should have the `anyOf` converted to a single value once this issue is fixed.

### PR DESCRIPTION
**Pull request by reidbaker**
#123643
- Add task to projects evaluated by flutter.gradle that will print the java version.
- Add integration test for the existence of javaVersion and the expected format. 
- Add gradle util to get the gradlew version for a specific platform (gradlew everywhere but windows). 

Why does this code need to exist? 
Figuring out what version of java is used by flutter/gradle is done in a few different ways that are not always aligned. 
See this issue https://github.com/flutter/flutter/issues/122609 ,
this issue https://github.com/flutter/flutter/issues/121501 this feature request https://github.com/flutter/flutter/issues/106416

As examples of why assuming the java version is dangerous. 
This task is code flutter can build upon and is the version gradle is using to build no matter how it is configured. 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/flutter_tools/gradle/flutter.gradle [hunk 1]
```diff
@@ -678,6 +678,17 @@ class FlutterPlugin implements Plugin<Project> {
         project.dependencies.add(configuration, dependency, config)
     }
 
+    // Add a task that can be called on flutter projects that prints the java version used in gradle.
+    // Format of the output of this task is used by other code in flutter to evaluate java version.
+    private static void addTaskForJavaVersion(Project project) {
+        // Warning: The name of this task is used by other code.
+        project.tasks.register('javaVersion') {
```
Comment (camsim99) [inline packages/flutter_tools/gradle/flutter.gradle:685]: How do you feel about calling it (print/show/get)JavaVersion? And is the warning for possible edits to the name?

**Review by camsim99 — APPROVED** [review]
LGTM!

**Comment by chunhtai** [comment]
I have a more general question: How will this task be used?

File: packages/flutter_tools/test/integration.shard/android_gradle_java_version.dart [hunk 2]
```diff
@@ -0,0 +1,55 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
```
Comment (chunhtai) [inline packages/flutter_tools/test/integration.shard/android_gradle_java_version.dart:1]: The name should end with *_test.dart

Comment (reidbaker) [reply packages/flutter_tools/test/integration.shard/android_gradle_java_version.dart:1]: good catch. fixed.

**Comment by reidbaker** [comment]
> I have a more general question: How will this task be used?

There are a handful of issues relating to identifying or configuring the java sdk. I see this a the source of truth for the version of java gradle is using. In practice I think it will be used to warn on incompatible java versions that other tools have. For example java 17 and gradle 6.7.

**Comment by gnprice** [comment]
Running a Gradle task is pretty slow. (Particularly due to #119196; I don't know how slow it would still be after fixing that.) So if you're thinking of a use case within a normal `flutter run` or `flutter build`, I think it'd be best to avoid having to run an extra Gradle task.

What should work to identify the Java version being used by Gradle is:
* The environment that the tool starts `gradlew` with is under the tool's control.
* The behavior of `gradlew` is that it looks at JAVA_HOME, and if it doesn't find an executable file at `"$JAVA_HOME/bin/java"` then it uses the `java` executable found in the PATH.
* So the tool can look at the JAVA_HOME and PATH environment variables it intends to pass to `gradlew`, and identify what `java` executable those point to.
* Then to find the version, one can run that `java` executable with the argument `-version`.

That approach isn't quite ground truth, because in principle the user could have an `android/gradlew` script that does something wild. But there a lot of ways that a user could in principle have made a mess, so it's impossible to be prepared for all of them. And I think this is a pretty unlikely one in practice — I think people generally avoid touching the gradlew scripts except by letting a Gradle upgrade do it.

In principle also a future Gradle version's `gradlew` could change that logic; but I think the behavior of "JAVA_HOME, then fall back to PATH" has been a widespread expectation in the Java world for a long time, and has been very stable in `gradlew`.

I think a task like this one could be useful as ground truth in a debugging situation, though. I'd just want to keep it out of a developer's normal `flutter run` or `flutter build` iteration loop.

**Comment by reidbaker** [comment]
> > In practice I think it will be used to warn on incompatible java versions that other tools have. For example java 17 and gradle 6.7.
> 
> In that situation, would this task be able to run? Or would it hit the same unhelpful error message as the main build tasks do?
> 
> My guess (though I don't actually know) would be that the error is happening before Gradle even gets to looking at a specific task, so that this task would hit the same unhelpful error message.

Thanks for these comments. I think you are correct that we will get an error before this code can run which makes this pr less valuable but not so much that I think we should drop this PR. I added a warning about speed and intended use and will not rely on this task for the other work I am doing around java versioning.

**Review by gnprice — APPROVED** [review]
LGTM with that plan

File: packages/flutter_tools/lib/src/android/gradle_utils.dart [hunk 3]
```diff
@@ -290,3 +289,12 @@ void exitWithNoSdkMessage() {
     'Try setting the ANDROID_SDK_ROOT environment variable.'
   );
 }
+
+// return gradlew file name based on the platform.
```
Comment (chunhtai) [inline packages/flutter_tools/lib/src/android/gradle_utils.dart:293]:
```suggestion
// Returns gradlew file name based on the platform.
```

**Review by chunhtai — APPROVED** [review]
LGTM

File: packages/flutter_tools/test/integration.shard/android_gradle_java_version_test.dart [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    result = await processManager.run(<String>[
+      './${getGradlewFileName(platform)}',
+      ...getLocalEngineArguments(),
+      '-q', // quiet output.
+      'javaVersion',
+    ], workingDirectory: androidApp.path);
+    // Verify that gradlew has a javaVersion task.
+    expect(result.exitCode, 0);
+    // Verify the format is a number on its own line.
+    expect(result.stdout.toString(), matches(RegExp(r'\d\n')));
```
Comment (christopherfujino) [inline packages/flutter_tools/test/integration.shard/android_gradle_java_version_test.dart:55]: 2 questions:

1. should we be matching 1 or more digits?
2. would this never be a `\r\n` on Windows?

File: packages/flutter_tools/test/general.shard/android/gradle_utils_test.dart [hunk 5]
```diff
@@ -9,6 +9,7 @@ import 'package:flutter_tools/src/base/logger.dart';
 import 'package:flutter_tools/src/base/platform.dart';
 import 'package:flutter_tools/src/cache.dart';
 import 'package:flutter_tools/src/project.dart';
+import '../../commands.shard/hermetic/build_windows_test.dart' as windows;
```
Comment (christopherfujino) [inline packages/flutter_tools/test/general.shard/android/gradle_utils_test.dart:12]: TIL about this check:

```
▌19:42:44▐ Test imports...
╔═╡ERROR╞═══════════════════════════════════════════════════════════════════════
║ The following file import a test directly. Test utilities should be in their own file.
║ /tmp/flutter sdk/packages/flutter_tools/test/general.shard/android/gradle_utils_test.dart
╚═══════════════════════════════════════════════════════════════════════════════
```

Comment (christopherfujino) [reply packages/flutter_tools/test/general.shard/android/gradle_utils_test.dart:12]: Failing the analysis shard

File: packages/flutter_tools/test/integration.shard/android_gradle_java_version_test.dart (hunk 4 above)
Comment (reidbaker) [reply packages/flutter_tools/test/integration.shard/android_gradle_java_version_test.dart:55]: 1. yes updated. 
2. Maybe I think I will update this to use the end of line sequence instead.

**Review by gmackall — APPROVED** [review]

**Pull request merged**